File: PrimeMinistersByJava/primeministers/Translator.java
package primeministers;

import java.lang.reflect.Constructor;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;
import java.awt.image.BufferedImage;
import libraries.Condition;
import libraries.ImageUtil;

public class Translator {

	private Table inputTable;

	private Table outputTable;

	public Translator(Object aClassAttribute) {
		Attributes.flushBaseDirectory();
		this.inputTable = new Table(aClassAttribute, new Attributes(aClassAttribute, "input"));
		this.outputTable = new Table(aClassAttribute, new Attributes(aClassAttribute, "output"));
	}

	/**
	* カレンダーを比較して差分を取得
	*/
	private int compareTo(Calendar aCalendar, Calendar anotherCalendar) {
		long differenceTime = aCalendar.getTimeInMillis() - anotherCalendar.getTimeInMillis();
		int millisOfDay = 86400000;
		return Math.abs((int)(differenceTime / millisOfDay)) + 1;
	}

	public Tuple computeNumberOfDays(Tuple aTuple) {
		String aDates = aTuple.values().get(this.inputTable.attributes().indexOfPeriod());
		String[] splittedDates = aDates.split("〜");
		String stringOfStartDate = splittedDates[0];
		Calendar startDate = this.getCalendar(splittedDates[0]);
		new Condition(() -> splittedDates.length >= 2)
		.ifThenElse(() -> {
			Calendar endDate = this.getCalendar(splittedDates[1]);
			aTuple.add(this.inputTable.attributes().indexOfPeriod() + 1, String.valueOf(this.compareTo(startDate, endDate)));
		},
		() -> {
			Calendar endDate = this.getCalendar();
			aTuple.add(this.inputTable.attributes().indexOfPeriod() + 1, String.valueOf(this.compareTo(startDate, endDate)));
		});

		return aTuple;
	}

	public void execute() {
		Downloader aDownloader = new Downloader(this.inputTable);
		List<Tuple> tuples = aDownloader.perform();
		Writer aWriter = new Writer(this.outputTable);
		AtomicInteger anAtomicInteger = new AtomicInteger();

		// ヘッダの情報をタプルに詰める
		List<String> header = new ArrayList<String>();
		this.outputTable.attributes().getEnum().forEach(header::add);
		System.out.println(header);
		tuples.add(0, new Tuple(header));


		tuples.forEach(aTuple -> {
			int index = anAtomicInteger.getAndIncrement();
			// HTMLタグをつける
			List<String> formattedList = new ArrayList<String>();
			Map<String, String> aMap = new HashMap<String, String>();

			new Condition(() -> index == 0)
			.ifThenElse(
				() -> {
					// Headerの処理
					aTuple.values().forEach(aColumn -> {
						formattedList.add(aWriter.tableHeader(aColumn));
					});
				},
				() -> {
					// Bodyの処理
					Tuple anotherTuple = this.computeNumberOfDays(aTuple);
					anotherTuple.values().forEach(aColumn -> {
						new Condition(() -> ImageUtil.isImageFile(aColumn))
						.ifThenElse(() -> {
							// imagefile.
							new Condition(() -> aColumn.startsWith("thumbnails"))
							.ifTrue(() -> {
								// thumbnails
								aMap.put("thumbnails", aColumn);
							});
							new Condition(() -> aColumn.startsWith("images"))
							.ifTrue(() -> {
								// images
								aMap.put("images", aColumn);
							});
							new Condition(() -> aMap.get("thumbnails") != null && aMap.get("images") != null)
							.ifTrue(() -> {
								// 画像をHTMLに充てる
								new Condition(() -> index % 2 == 0)
								.ifThenElse(() -> {
									// even.
									formattedList.add(aWriter.getImageWithYellow(aMap));
								},
								() -> {
									// odd.
									formattedList.add(aWriter.getImageWithBlue(aMap));
								});
							});
						},
						() -> {
							// not imagefile.
							// is anAtomicInteger even.
							new Condition(() -> index % 2 == 0)
							.ifThenElse(() -> {
								// even.
								formattedList.add(aWriter.getTableDataWithYellow(aColumn));
							},
							() -> {
								// odd.
								formattedList.add(aWriter.getTableDataWithBlue(aColumn));
							});
						});
					});
				});
			this.outputTable.add(new Tuple(formattedList));
		});
		aWriter.write(this.outputTable);
	}

	private Calendar getCalendar() {
		Calendar instanceOfCalendar = Calendar.getInstance(Locale.JAPAN);
		String thisYear = Integer.toString(instanceOfCalendar.get(Calendar.YEAR)).concat("年");
		String thisMonth = Integer.toString(instanceOfCalendar.get(Calendar.MONTH) + 1).concat("月");
		String thisDate = Integer.toString(instanceOfCalendar.get(Calendar.DATE)).concat("日");
		String stringOfEndDate = thisYear.concat(thisMonth.concat(thisDate));
		return this.getCalendar(stringOfEndDate);
	}

	private Calendar getCalendar(String aString) {
		Calendar aCalendar = Calendar.getInstance();
		try {
			String[] firstSplit = aString.split("年");
			String[] secondSplit = firstSplit[1].split("月");
			String[] thirdSplit = secondSplit[1].split("日");
			int year = 1990;
			int month = 1;
			int day = 1;
			year = Integer.valueOf(firstSplit[0]);
			month = Integer.valueOf(secondSplit[0]) - 1;
			day = Integer.valueOf(thirdSplit[0]);
			aCalendar.set(year, month, day);
		} catch(ArrayIndexOutOfBoundsException | NumberFormatException e) {
			e.printStackTrace();
		}
		return aCalendar;
	}

}

File: PrimeMinistersByPython/primeministers/translator.py
from .attributes import Attributes
from .table import Table
from .downloader import Downloader
from .tuple import Tuple
from .writer import Writer

import datetime
import imghdr
import re as regular_expression
import os
import sys
sys.path.append('../')
from libraries.condition import Condition

class Translator:
	def __init__(self, a_class_name):
		Attributes.flush_base_directory()
		self.input_table = Table(a_class_name, Attributes(a_class_name, "input"))
		self.output_table = Table(a_class_name, Attributes(a_class_name, "output"))
		return

	def compute_number_of_days(self, a_tuple):
		index_of_period = self.input_table.get_attributes().index_of_period()
		dates = a_tuple.get_values()[index_of_period].split('〜')
		get_calendar = lambda a_bool, a_date: list(filter(a_bool, regular_expression.split('[年 | 月 | 日]', a_date)))
		splitted_start_dates = list(map(lambda a_string: int(a_string), get_calendar(lambda a_string: a_string != '', a_date = dates[0])))
		start_date = datetime.datetime(year=splitted_start_dates[0], month = splitted_start_dates[1], day=splitted_start_dates[2])
		Condition(dates[-1] != '').if_then_else(
			lambda: a_tuple.add(index_of_period + 1, (self.get_datetime(dates, get_calendar) - start_date).days + 1),
			lambda: a_tuple.add(index_of_period + 1, (datetime.datetime.today() - start_date).days + 1)
		)
		return a_tuple

	def execute(self):
		a_downloader = Downloader(self.input_table)
		tuples = a_downloader.perform()
		header = [an_item.name for an_item in self.output_table.get_attributes().get_enum()]
		tuples.insert(0, Tuple(header))
		a_writer = Writer(self.output_table)
		is_image = lambda a_column: str(a_column).startswith('images')
		is_thumbnail = lambda a_column: str(a_column).startswith('thumbnails')
		for an_index, a_tuple in enumerate(tuples):
			formatted_list = []
			a_dictionary = {}
			Condition(an_index == 0).if_then_else(
				lambda: [ formatted_list.append(a_writer.table_header(a_string)) for a_string in a_tuple.get_values() ],
				lambda: [ Condition(
					is_image(a_column) or is_thumbnail(a_column)
					).if_then_else(
					lambda: [
						Condition(is_image(a_column)).if_true(
							lambda: a_dictionary.update({"images": a_column})
						),
						Condition(is_thumbnail(a_column)).if_true(
							lambda: a_dictionary.update({"thumbnails": a_column})
						),
						Condition('thumbnails' in a_dictionary and 'images' in a_dictionary).if_true(
							lambda: Condition(an_index % 2 == 0).if_then_else(
								lambda: formatted_list.append(a_writer.get_image_with_yellow(a_dictionary)),
								lambda: formatted_list.append(a_writer.get_image_with_blue(a_dictionary))
							)
						)
						],
					lambda: Condition(an_index % 2 == 0).if_then_else(
						lambda: formatted_list.append(a_writer.get_table_data_with_yellow(a_column)),
						lambda: formatted_list.append(a_writer.get_table_data_with_blue(a_column))
					)
				) for a_column in self.compute_number_of_days(a_tuple).get_values() ]
			)
			self.output_table.add(Tuple(formatted_list))
		a_writer.write(self.output_table)
		return
	
	def get_datetime(self, dates, get_calendar):
		year, month, day = tuple(map(lambda a_string: int(a_string), get_calendar(lambda a_string: a_string != '', dates[1])))
		return datetime.datetime(year, month, day)

	def make_table_header(self, formatted_list, a_tuple):
		for a_string in a_tuple.get_values():
			formatted_list.append(a_writer.table_header(a_string))
		return formatted_list